### PR TITLE
Rust cc log support

### PR DESCRIPTION
**Pull request by slyphon**
Add a binding layer between rust's `log` crate and `cc_log`'s logging facilities. This is intended to allow rust to log using its own facilities, but have the messages routed through cc_log.

Included are integration tests from both the rust and C sides, plus a backport of the bstring wiring from the pelikan repo.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -143,15 +148,21 @@ endif(CHECK_FOUND)
...
 find_package(Threads)
 
-enable_language(Rust)
-include(CMakeCargo)
-
 
 # where to find include files
-include_directories(${include_directories}
+include_directories(
+    ${include_directories}
```
Comment (thinkingfish) [inline CMakeLists.txt:154]: do you want to prepend instead of append?

File: include/rust/cc_log_rs.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
+/* ccommon - a cache common library.
+ * Copyright (C) 2013 Twitter, Inc.
```
Comment (thinkingfish) [inline include/rust/cc_log_rs.h:2]: at some point we need to fix the year line across the repo, but not today

File: rust/ccommon_rs/src/log.rs [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+fn make_logger() -> &'static Logger {
+    // NOTE(simms): this is how you get a &'static T reference.
+    // This drove me nuts for hours trying to figure out:
+    // * A static reference is a reference that lives for the life of the program
+    // * If you create a new Box, copy an object to it, then convert that to a raw
+    //   pointer, you've essentially malloced, then forgotten (in terms of rust lifetimes)
+    //   that object. This essentially makes that reference _live for the life of the program_.
+    // * There is an implicit conversion (bijective) between &mut T and *mut T, so we
+    //   can assign a *mut T to a variable that expects a &'static mut T.
+    Box::leak(Box::new(Logger::new()))
```
Comment (thinkingfish) [inline rust/ccommon_rs/src/log.rs:178]: 😱

File: rust/ccommon_rs/src/log.rs [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+/// [`log_rs_setup`]: fn.log_rs_setup.html
+/// [`LoggerNotSetupError`]: enum.LoggerStatus.html
+/// [`LoggerAlreadySetError`]: enums.LoggerStatus.html
+///
+/// # Undefined Behavior
+///
+/// If the `logger` pointer becomes invalid before [`log_rs_unset`] is called, the
+/// behavior is undefined.
+///
+/// [`log_rs_unset`]: fn.log_rs_unset.html
```
Comment (thinkingfish) [inline rust/ccommon_rs/src/log.rs:344]: I lean toward either naming the rs-C APIs `rs_*` (e.g. `rs_log_setup()`) or `*_rs`(`log_setup_rs()`) even just for the sake of simpler regex for search.

File: src/cc_log.c [hunk 5]
```diff
@@ -37,6 +38,30 @@
 static log_metrics_st *log_metrics = NULL;
 static bool log_init = false;
 
+/* this function is called from rust so that it can use log_setup */
+log_metrics_st *
+log_metrics_create()
```
Comment (thinkingfish) [inline src/cc_log.c:43]: `()`->`(void)`

**Comment by slyphon** [comment]
Gonna close this, as I'm working on breaking it into single and multithreaded implementations

**Pull request closed without merging**